src/auction.js

```diff
@@ -514,28 +514,67 @@ function tryAddVideoBid(auctionInstance, bidResponse, afterBidAdded, {index = au
   }
 }
 
-export const callPrebidCache = hook('async', function(auctionInstance, bidResponse, afterBidAdded, videoMediaType) {
-  store([bidResponse], function (error, cacheIds) {
-    if (error) {
-      logWarn(`Failed to save to the video cache: ${error}. Video bid must be discarded.`);
-
-      doCallbacksIfTimedout(auctionInstance, bidResponse);
-    } else {
-      if (cacheIds[0].uuid === '') {
-        logWarn(`Supplied video cache key was already in use by Prebid Cache; caching attempt was rejected. Video bid must be discarded.`);
+const storeInCache = (batch) => {
+  store(batch.map(entry => entry.bidResponse), function (error, cacheIds) {
+    cacheIds.forEach((cacheId, i) => {
+      const { auctionInstance, bidResponse, afterBidAdded } = batch[i];
+      if (error) {
+        logWarn(`Failed to save to the video cache: ${error}. Video bid must be discarded.`);
 
         doCallbacksIfTimedout(auctionInstance, bidResponse);
       } else {
-        bidResponse.videoCacheKey = cacheIds[0].uuid;
+        if (cacheId.uuid === '') {
+          logWarn(`Supplied video cache key was already in use by Prebid Cache; caching attempt was rejected. Video bid must be discarded.`);
 
-        if (!bidResponse.vastUrl) {
-          bidResponse.vastUrl = getCacheUrl(bidResponse.videoCacheKey);
+          doCallbacksIfTimedout(auctionInstance, bidResponse);
+        } else {
+          bidResponse.videoCacheKey = cacheId.uuid;
+
+          if (!bidResponse.vastUrl) {
+            bidResponse.vastUrl = getCacheUrl(bidResponse.videoCacheKey);
+          }
+          addBidToAuction(auctionInstance, bidResponse);
+          afterBidAdded();
         }
-        addBidToAuction(auctionInstance, bidResponse);
-        afterBidAdded();
       }
-    }
+    });
   });
+};
+
+const batchAndStore = (() => {
+  let batches = [[]];
+  let debouncing = false;
+
+  return function(auctionInstance, bidResponse, afterBidAdded) {
+    const batchSize =
+      typeof config.getConfig('cache.batchSize') === 'number' && config.getConfig('cache.batchSize') > 0
+        ? config.getConfig('cache.batchSize')
+        : 1;
+
+    const batchTimeout =
+      typeof config.getConfig('cache.batchTimeout') === 'number' && config.getConfig('cache.batchTimeout') > 0
+        ? config.getConfig('cache.batchTimeout')
+        : 0;
+
+    if (batches[batches.length - 1].length >= batchSize) {
+      batches.push([]);
+    }
+
+    batches[batches.length - 1].push({auctionInstance, bidResponse, afterBidAdded});
+
+    if (!debouncing) {
+      debouncing = true;
+      setTimeout(() => {
+        batches.forEach(batch => storeInCache(batch));
+        batches = [[]];
+        debouncing = false;
+      }, batchTimeout);
+    }
+  }
+})();
+
+export const callPrebidCache = hook('async', function(auctionInstance, bidResponse, afterBidAdded, videoMediaType) {
+  batchAndStore(auctionInstance, bidResponse, afterBidAdded);
```

Comment:
It's because here you always call the batched version, which means the cache will always be delayed. (Even if the batchTimeout is 0, it still waits for the microtask queue to clear before running it; which means that all tests that expect that to have happened already, fail). Plus there's a bunch of red herring failures because they're not configured that well and can't deal with some previous setup / teardown not running.


if you change line 577 to `storeInCache([{auctionInstance, bidResponse, afterBidAdded}])` all tests pass, which is still not ideal because it means that the batching logic is untested. An alternative would be to change line 567 to only use setTimeout if batchTimeout is greater than zero.

Something extra I should mention - I don't know how the cache endpoint works, so I am expecting it to do what your code assumes it does  -  which is always return a cacheId for each item that was requested, in the same order. Does that sound correct to you?

Reply:
In regards to the cache endpoint, I was wondering the same thing.  According to the README page here, https://github.com/prebid/prebid-cache, the sample responses in the sections titled `Puts Parameters` and `Overwritting Values` look to demonstrate that responses do come back in the same order.

However, I also found these lines in the test file for cache endpoint:

without custom keys:

- https://github.com/prebid/prebid-cache/blob/master/endpoints/put_test.go#L150 (looks like the same elements from the response and request are checked for here, not 100% sure if the order is preserved in this scenario though..)

with custom keys

- https://github.com/prebid/prebid-cache/blob/master/endpoints/put_test.go#L176 (looks like order is preserved here 👍 )


Do you know anyone we could ask to confirm if the ordering will work as expected or if a change is needed?

Reply:
I asked in the prebid-server slack channel and I got confirmation that the order matches for the Go version, and no reply for the Java version.

Reply:
Oh nice, that's good to know for the Go version!  Hopefully the same outcome for the Java one too..